### PR TITLE
feat: make content models connection agnostic

### PR DESCRIPTION
**Pull request by scottdover**
**Summary**
This PR is a prerequisite  for integrating sas server into the extension (see overall design, here: https://github.com/sassoftware/vscode-sas-extension/issues/889). This makes a few additional refactors outside of those mentioned in https://github.com/sassoftware/vscode-sas-extension/issues/889.

Favorites. Previously, we were storing a favorite id w/n `ContentItem.flags`. ~This change delays finding that favorite id _until_ we're actually doing something with that favorite.~ Now, we are storing the uri to simplify the deletion process. 

Notebook -> flow conversion. When we first implemented this change, there was some question around whether this belong in content data provider or not. In this PR, functionality is moved to a `NotebookToFlowConverter` which makes use of a `ContentModel` directly.

It's also worth mentioning that this moves some additional properties into `ContentItem` that were previously utils functions, and updates _every_ content adapter method to return a content item decorated with those new properties. Additionally, this splits utils such that `ContentNavigator/utils` has things that are connection agnostic, while `rest/utils` contains things that are connection specific.

**Testing**
 - [x] Test creating file
 - [x] Test creating folder
 - [x] Test moving files & folders
 - [x] Test recycling files & folders
 - [x] Test deleting files & folders
 - [x] Test add to favorites for files and folders
 - [x] Test remove from favorites for files and folders
 - [x] Test editing files (regular files & favorites & favorites w/n folders)
 - [x] Test downloading files and folders
 - [x] Test converting sas notebooks to flows (test in explorer as well as sas content)
 - [x] Test renaming files & folders
 - [x] Test emptying recyling bin
 - [x] Test refreshing content
 - [x] Test filename statement
 - [x] Test fetching viya cadence
 - [x] Test uploading files (context menu, dragging from file system, dragging from explorer)
 - [x] Test collapsing all files
 - [x] Make sure multi-select is working the same on this branch and main
 - [x] Test removing a favorited item from 'my folder'

**Todos**
 - [x] Address test coverage/build issues

**Observations**
 - Unable to download flows. This is because flows are stored at a different endpoint than regular files

### CODE REVIEW

File: client/src/connection/rest/SASContentAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+
+    async function getTypeDefinition(fileName: string): Promise<string> {
+      const defaultContentType = "file";
+      const ext = fileName.split(".").pop().toLowerCase();
+      if (ext === "sas") {
+        return "programFile";
+      }
+
+      try {
+        const typeResponse = await this.connection.get(
```
Comment (scottdover) [inline client/src/connection/rest/SASContentAdapter.ts:458]: Make sure this works

File: client/src/connection/rest/SASContentAdapter.ts [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+      return false;
+    }
+    return true;
+
+    async function getDeleteMemberUri(): Promise<string> {
+      if (item.flags?.isInMyFavorites) {
+        return getLink(item.links, "DELETE", "delete")?.uri;
+      }
+
+      const myFavoritesFolder = this.getDelegateFolder("@myFavorites");
```
Comment (scottdover) [inline client/src/connection/rest/SASContentAdapter.ts:565]: Perhaps this does work?

**Comment by scottdover** [comment]
Hey @smorrisj. I made some changes matching the general design of https://github.com/sassoftware/vscode-sas-extension/issues/889.

That said, after these changes, `ContentModel` looks pretty slim and serves no _real_ purpose of its own other than:
 - Presenting some error messaging back to content data provider
 - providing a different data format (i.e. string vs buffer)

After working through this, it almost feels like `ContentModel` should be an abstract class that connection-specific things extend. Thoughts?

**Comment by scottdover** [comment]
@scnwwu wondering if you have any thoughts about the above?

**Comment by scnwwu** [comment]
> @scnwwu wondering if you have any thoughts about the above?

defer to @boyce-w who is more familiar with the Content part.

**Review by boyce-w — APPROVED** [review]

**Comment by Zhirong2022** [comment]
ContextMenu name changed from 'Remove from My Favorites' to 'Add to My Favorites' if the file already has been added to My Favorites.
1.Create a file under My Folder
2.Add the file to My Favorites through ContextMenu
3.Open ContextMenu against the file again
image

**Comment by Zhirong2022** [comment]
Multi-select is not working the same on this branch and main.
Multi-selection through pressing Shift or Ctrl works well on this branch. It does not work well on main.

**Comment by Zhirong2022** [comment]
Multi-select some files and perform 'Add to My Favorites' through context menu, it does not move all files 
image

**Comment by scottdover** [comment]
> ContextMenu name changed from 'Remove from My Favorites' to 'Add to My Favorites' if the file already has been added to My Favorites. 1.Create a file under My Folder 2.Add the file to My Favorites through ContextMenu 3.Open ContextMenu against the file again image

> Multi-select some files and perform 'Add to My Favorites' through context menu, it does not move all files image

It looks like this second one is an issue on main as well. Both of these have been resolved in https://github.com/sassoftware/vscode-sas-extension/pull/1079/commits/da0417b33d5994c55590fc5b73c36259c884ce65 .

> Multi-select is not working the same on this branch and main. Multi-selection through pressing Shift or Ctrl works well on this branch. It does not work well on main.

Is this still an issue for you, @Zhirong2022 . I noticed this problem on main and my branch several weeks ago when implementing this, but it since seems to have resolved itself (I'm guessing it may've been a vscode bug)

**Comment by Zhirong2022** [comment]
@scottdover , thank you for your help.  I upgraded my VSC to the latest version, multi-selection works well through pressing Shift or Control.  And other issues have been fixed in the branch package.

**Review by Zhirong2022 — APPROVED** [review]

**Pull request merged**